### ISSUE TITLE
Gtk4: any plans?

### ISSUE THREAD

**Issue by lytico** [GtkSharp/GtkSharp#166]
are there any plans to support GTK4?

**Comment by harry-cpp** [comment]
Yes, tho I will probably break the compatibility and write a brand new generator.

**Comment by gavr123456789** [comment]
OMG, I'm so happy to hear this, I thought with the death of the monodevelope project in 2018, there will be no new GTK#.  I look forward to.

**Comment by mikkeljohnsen** [comment]
@harry-cpp What would it take to make a joined afford between GLibSharp/GtkSharp and this one ?

I also would like to see GStreamerSharp to be able to settle on using the joined version.

Have you started the wrap for Gtk4 ?

**Comment by harry-cpp** [comment]
> What would it take to make a joined afford between GLibSharp/GtkSharp and this one ?

I originally split out from that repo as the other 2 guys weren't C# users so we had a bit heated discussions about using .csproj project files, using nugets, etc.

> I also would like to see GStreamerSharp to be able to settle on using the joined version.

I could start maintaining a GStreamerSharp in this repo, it would not be too hard.

> Have you started the wrap for Gtk4 ?

I have started writing my own wrapper but haven't touched it in a while.

**Comment by mikkeljohnsen** [comment]
@harry-cpp I really just want more eyes on the same wrapper.

We are struggling with memory problems and GStraemerSharp seems to not want to choose this wrapper because of memory problems (something about floating refs). So if everyone could agree on one wrapper, it would be easier to solve problems.

**Comment by BehroozAmoozad** [comment]
Please shoot me a message when the work on Gtk4 starts, I'd be very interested to help out.   
I have experience with gtk 2 and 3, Am already up to speed with some of changes in gtk4, and been using this library and read the interesting parts of the source code since it came out. Also have 10+ years of experiece writing Linux software using mostly mono and monodevelop. and ~15 for windows.   
I'm sure you can use me.

**Comment by badcel** [comment]
@harry-cpp When we discussed the [gir.core project](https://github.com/gircore/gir.core) in discord I wanted to try to generate a C# wrapper for GTK which focuses more on a nice API and kind of hides GTK from the user. Back then you recommended to me to create a real wrapper over GTK and replicating the GTK API via a generator.

In the meantime we are now 3 people working actively on this and we switched directions and actually try to do what you suggested. There were a lot of discussions and we try to generate a modern C# wrapper for GObject Introspection in general (GTK beeing only one part).

This is a whole lot of work and any support is welcome as we are still in the early phase (especially after shifting directions, which made the things harder because we need to deeply integrate with GObject and GTK).

We still write most of the bindings manually but want to get rid of this. We actually do it to see where some potential pitfalls could be as we want to rewrite the current generator completely.

I think the general direction of both projects is similar now and perhaps it is worthwile to see what we have done so far.

**Comment by harry-cpp** [comment]
Uuu. I'll check it out during the weekend.

**Comment by harry-cpp** [comment]
And I forgot XD

Anyhow now that Gtk 4 is officially out I'll regenerate the api files in a week or two and try to support it with the old fashioned method for the time being. Need to also configure a separate branch for the Gtk 3 version as I can see people wanting to stick with it.

As for my thoughts about gir.core, I hope it can supersede this project, I currently don't have much time to dig into its code, or help out at all for that matter.

**Comment by lytico** [comment]
see also: https://github.com/GtkSharp/GtkSharp/issues/213#issue-790778747

**Comment by litch0** [comment]
I made a project with a simple interop to test if it was Ok and promessing to make gtk4 support with rust, I suprise myself because this is the first project that i don't have to fight with cargo for it to work. [LinkToIt](https://github.com/litch0/rust-chsarp-interop) I am planning on starting the support for it, but It will take me sometime to actually make some progress because I have a lot of school work todo, But in the vacation I am planning on starting.

**Comment by trampster** [comment]
There is some discussion happening about the future of .net Maui linux support on here https://github.com/jsuarezruiz/forms-gtk-progress/issues/31

There seems to be interest in using GTK4, and they are asking for feedback. Would you guys be interested in working with microsoft on .net maui linux support?

**Comment by gavr123456789** [comment]
@trampster looks like better to ask it there https://github.com/gircore/gir.core, but as I see @badcel already here

**Comment by badcel** [comment]
@gavr123456789 thanks for the hint I completely missed the discussion.

**Comment by litch0** [comment]
I started the C P/Invoke of the Gtk libraries, soon I will start the abstraction, when that's done, I will make a new repository, because the APIs are completely different

**Comment by mikkeljohnsen** [comment]
Is there any progress in the port fot Gtk4 ?

I would like to start testing it.

**Comment by harry-cpp** [comment]
I've not started work on it yet, I thought I would find time, but my day job kept me drained during the week days, and I've dedicated weekends to my Super Secret Project™.

**Comment by BehroozAmoozad** [comment]
Same here. although my super secret and not so super secret projects will end in a month or two and I can dedicate my own time and another developer familiar with gtk3 and c# to this. If you have the time to split it up into tasks (e.g. write all Interops that involve X like Y) I think I may be able to do at least parts of it before I'm actually free to be on this 24/7.

**Comment by litch0** [comment]
I have start porting, but I only have the minimum to open a window with buttons, it's just to much functions to write

**Comment by litch0** [comment]
Also, when I say minimum I really mean it, you have to call native methods directly.
```c#

IntPtr app = GtkNativeMethods.gtk_application_new("org.app.id", 0);

```

I realized that it would be easier to rewrite the the gtk library rather adapt the existing one, because, in my understanding of their publications, it has a lot of breaking changes

**Comment by litch0** [comment]
But if you are interested, here is the repo https://gitlab.com/gtksharp4/gksharp4

**Comment by harry-cpp** [comment]
Thats not porting, thats writing a brand new bindings, if you wanna do that, you better help out the git.core team, those guys will eventually have bindings which are much better than the ones here: https://github.com/gircore/gir.core

**Comment by seikosantana** [comment]
> But if you are interested, here is the repo https://gitlab.com/gtksharp4/gksharp4

It seems the repo is no longer there?

**Comment by litch0** [comment]
No, I just had written a wrapper by hand to the c library and realize that my project was stupid because of that. You would need to write all the functions that were missing (like label, VBox, button, etc) just to get an app running. On top of that gir already has a generator so you could just use them, it's the same thing but my it's written by hand. So I deleted it

**Comment by BehroozAmoozad** [comment]
Well, I have good news. My thing worked, tad better than I initially expected.  
I have successfully converted ~1M SLOC of C# 10 code to vala with very few errors. Just working on kinks now. 
I will probably release it as a SaaS for commercial projects and free for OSS projects. It will probably take me a few more weeks to completely polish it enough to do more than a single-shot run on my own code.

**Comment by litch0** [comment]
Good luck

**Comment by litch0** [comment]
> Well, I have good news. My thing worked, tad better than I initially expected.
> I have successfully converted ~1M SLOC of C# 10 code to vala with very few errors. Just working on kinks now.
> I will probably release it as a paid code conversion service for commercial projects and free for OSS projects. It will probably take me a few more weeks to completely polish it enough to do more than a single-shot run on my own code.

Good luck @BehroozAmoozad

**Comment by seikosantana** [comment]
> No, I just had written a wrapper by hand to the c library and realize that my project was stupid because of that. You would need to write all the functions that were missing (like label, VBox, button, etc) just to get an app running. On top of that gir already has a generator so you could just use them, it's the same thing but my it's written by hand. So I deleted it

Does that mean we can use the generated GtkSharp 4 from gir instead of having to work with MAUI?

**Comment by litch0** [comment]
Right now there is only https://github.com/gircore/gir.core/tree/develop/src/Libs/Gtk-3.0/Classes working, so If you want to work with something more specific you would have to work with the C library. Therefore, I don't think that It would be a great Idea to work with it right now. There isn't any classes for Gtk4 only 3, so you would have to work with pointers, unsafe conversions and a some other incompatibilities like structures.

**Comment by seikosantana** [comment]
> Right now there is only https://github.com/gircore/gir.core/tree/develop/src/Libs/Gtk-3.0/Classes working, so If you want to work with something more specific you would have to work with the C library. Therefore, I don't think that It would be a great Idea to work with it right now. There isn't any classes for Gtk4 only 3, so you would have to work with pointers, unsafe conversions and a some other incompatibilities like structures.

I see. Thanks. Maybe I'll find some other alternatives for now, if any. Sticking with GTK3 is not a problem, but it seems some things doesn't get fixed anytime soon.

**Comment by litch0** [comment]
yeah

**Comment by ErickHoward** [comment]
Are there any updates on the migration to GTK4?

**Comment by BehroozAmoozad** [comment]
Yes. Since everyone was just running away from working on this I decided to do it myself. And since microsoft outright lied about Gtk support in MAUI I decided I will never trust microsoft again decided to write a C# to Vala convertor instead of wasting my time on helping microsoft's EEE.  
I'll create the repo in a couple weeks and upload the first version.
Right now the converter converts extremely large codebases without any syntax errors. and deals with the shenanigans in overloaded constructors, events, partial classes, interfaces, delegates, exceptions, is/as, and a few other things.  
Good riddance, microsoft.

**Comment by ErickHoward** [comment]
@BehroozAmoozad wow that sounds amazing! Reply here when you release it! I'll look forward to it.

**Comment by lmtr0** [comment]
@BehroozAmoozad  oh thats nice, I  would love to help however I'm finishing my IB diploma so I don't have time for anything.

**Comment by seikosantana** [comment]
I prefer having GtkSharp that targets GTK4 instead of working with MAUI that targets GTK4. And I read about Vala and some reasons why it isn't recommended for new projects so I wonder if it really worth the time to invest some time learning Vala for writing GTK applications.

**Comment by nlogozzo** [comment]
What's the status on this?

**Comment by BehroozAmoozad** [comment]
Well, it works good enough for most syntax but there was still a few problems with `IComparable<T> : IComparable`,  `IEnumerable<T> : IEnumerable`, `ILst<T> : IList`, etc which led me to re-implement the .net stdlib and do a lot of find/replace and refactoring to remove the non-generic versions which has worked for the most part with minimal code rewrite.  
The big thing left is the `T` return values in generic methods that can't be  larger than 32-bit value types due to historical reasons in glib and C. The codegenerator should be completely usable once that's fixed or I find a work-around for it.

**Comment by BehroozAmoozad** [comment]
> I prefer having GtkSharp that targets GTK4 instead of working with MAUI that targets GTK4. And I read about Vala and some reasons why it isn't recommended for new projects so I wonder if it really worth the time to invest some time learning Vala for writing GTK applications.

Vala has come a long way since then and now supports a lot of things it didn't then.

**Comment by polyjitter** [comment]
CSharp -> Vala converter is kinda missing the point :/ Vala exists, sure - it doesn't have the backing of dotnet core, however, and it doesn't let me use F# for functional programming.

**Comment by RetroSeven** [comment]
This would greatly contribute to .NET having both MAUI **as well as** native development support for Windows, Linux, macOS, iOS and Android, which is definitely something I'd be thrilled about.

**Comment by BehroozAmoozad** [comment]
The .net to vala project got scrubbed. because vala doesn't support longer than 32 bit gobject pointers or some weird stuff I no longer even want to remember.  
It took a good few months of effort until we hit that wall. no way around that.

**Comment by RetroSeven** [comment]
> And since microsoft outright lied about Gtk support in MAUI I decided I will never trust microsoft again
You can just write it yourself instead of getting mad at them. You don't need a _Microsoft Seal of Approval_ on it. When this library gets updated to support Gtk4, you can just make an adapter for MAUI yourself, or use this library directly.
Much easier than a C# to Vala converter that would be barely usable with the existing and very versatile .NET infrastructure, **which is the exact reason why someone would want to use C# to begin with.**

**Comment by BehroozAmoozad** [comment]
I've been using Gtk# full time for at least 6 years and .net for at least 17. and I'm fed up with it. It's not a decision I took lightly to convert millions of lines of code. It's worth it to be rid of C# and all the bs that follows. Everything on that list is a reason why I hate .net. System.Reflection is slow AF and if you use it to do anything it only means your code smells and absolutely nothing else . VB.NET is a franken language, IronPython is worst of both worlds. F# is ok if you don't know haskell or rust or erlang. Every single time that I want to do something with .net that might truly make a difference I hit something that they didn't care to implement or allow. there are like a gazillion private classes in .net that they do not let you inherit or override. the roslyn code generator is a joke. the roslyn syntaxtree is an even worse joke. the ISyntaxModel is so bad it looks like it was written in rush for a PoC.
I'm 100% making an emotional decision here. They have lied. They have obstructed progress. They have deliberately removed support. They have Intentionally removed features to make .net unusable on other platforms. They have on 3 instances that I have personal knowledge of, performed EEE. I NO LONGER WANT TO WORK WITH .NET. 
If you love .net so much stick with it. if you've worked with it for as long as I have and have hit the walls that I have then you understand my reasons.

**Comment by polyjitter** [comment]
Then don't use GtkSharp. I don't know why you're still here.

**Comment by BehroozAmoozad** [comment]
Because I made this mistake a few years back and am stuck with it, should
be as good a reason as any?

**Comment by RetroSeven** [comment]
Well then, it is your mistake and not ours. We just want Gtk4 in C# because we like C# and we like .NET. If you are unwilling to use a community developed package for MAUI on Linux, don't use MAUI and don't use .NET.

**Comment by hhyyrylainen** [comment]
I ended up using [Avalonia](https://github.com/AvaloniaUI/Avalonia), because it's fully supported by the developers for all desktop platforms (showing they actually care about being crossplatform). It does still have some pretty rough edges like missing tutorials, bugs and a few features I would have liked to see. But after looking at a bunch of C# GUI frameworks and testing out a couple it was the most crossplatform and had the fewest serious problems I ran into in the first few hours of testing it out.

**Comment by HubKing** [comment]
Is this project abandoned? I mean, the last update was about 2 years ago. A lot has changed since then; GTK4 is no longer a novelty, and other languages like Python, Java, Rust, even bloody JavaScript, now have GTK4 libraries. Is there any rough release plan or schedule for C# GTK4?

**Comment by seikosantana** [comment]
Seems like the Gir Core now has partially support for GTK4 while in early stage https://github.com/gircore/gir.core

**Comment by nlogozzo** [comment]
> Seems like the Gir Core now has partially support for GTK4 while in early stage https://github.com/gircore/gir.core

I wouldn't call it partial support. It's really well done. Has support for both gtk4 and libadwaita. It's what I use to develop Denaro :)

**Comment by HubKing** [comment]
> > Seems like the Gir Core now has partially support for GTK4 while in early stage https://github.com/gircore/gir.core
> 
> I wouldn't call it partial support. It's really well done. Has support for both gtk4 and libadwaita. It's what I use to develop Denaro :)

What exactly is Git.Core? In the "About" section it says that it is "A C# binding **generator** for GObject", but seeing the table, it seems that it comprises a GTK library, like GtkSharp. Seeing "nlogozzo"'s reply, it looks more like the latter...

Also, what does "support: partial" mean exactly? Is it at least as usable as GtkSharp? I can use common GTK4 widgets?

**Comment by seikosantana** [comment]
> I wouldn't call it partial support
Hmmm I haven't tried the GTK4 binding of Gir Core so I can't really tell, but the pages wrote partial support. But i do see the demo GTK4 application somewhere so i think it is actually usable.

> What exactly is Gir.Core?
I'm no expert in GTK and Gir stuff, but if I understand correctly it is a tool to generate C# binding libraries out of the GObject based (native) libraries, so we can consume the generated C# library and write, for example GTK4 application, and this GTK4 library is generated by Gir.Core, instead of being written by hand. Correct me if I'm wrong tho

**Comment by nlogozzo** [comment]
gir.core is a C# language binding for GTK and its libraries. It is built from generators that create the C# code from gobject introspection data, as opposed to manually creating `DllImports` for everything. 
Therefore, since gtk4, libadwaita, gdk, etc... all have gobject introspection data, they are all supported by this library.

Right now, most libraries are marked as partial support since the generator isn't perfect. 
There are still some random functions missing, for example: `gtk_css_provider_load_from_data`, and working with some structs such as `GDateTime` needs to be done manually via DllImport. 
But other than those minor cases, which will be fixed in the next release coming soon, it pretty much supports everything. 

As I said, I use it to build [Denaro](https://github.com/nlogozzo/NickvisionMoney) and I also have a [minimal template](https://github.com/nlogozzo/NickvisionApplication) that can help you get started. 
Feel free to ask me any questions you have or any problems you may run into :)

**Comment by HubKing** [comment]
> As I said, I use it to build [Denaro](https://github.com/nlogozzo/NickvisionMoney) 

It seems that it uses WinUI for Windows and GTK4 for Linux. Is this because Gir.Core + GTK4 does not work on Windows, or is it just because you wanted to use native Windows UI? I mean, if possible, I do not want to do double amount of work and just use GTK4 for all desktop OS.

**Comment by nlogozzo** [comment]
> > As I said, I use it to build [Denaro](https://github.com/nlogozzo/NickvisionMoney) 
> 
> It seems that it uses WinUI for Windows and GTK4 for Linux. Is this because Gir.Core + GTK4 does not work on Windows, or is it just because you wanted to use native Windows UI? I mean, if possible, I do not want to do double amount of work and just use GTK4 for all desktop OS.
> 

I separated into 2 versions as I wanted the Windows version to look native, but using GTK 4 for Windows absolutely works! 

You just have to make sure all of the gtk dependencies are installed correctly on Windows for anyone who uses your app...which can be a little bit of a pain for inexperienced users.

I believe future versions of gir.core will have all dependencies bundled into the library to remove the need for extra setup.

But for now, there are some instructions on the gir.core webpage but feel free to join the Matrix room and talk to the developers about getting it setup on Windows...they'll be glad to help :)

**Comment by Soufleuse** [comment]
Hi, I believe to understand Gtk4 would work on Windows. Is there any installer (msi/exe) available for that?

Let me know if it exists.

**Comment by lytico** [comment]
@Soufleuse  GtkSharp does not support Gkt4.
If you look for Gtk4-Support, you can try https://github.com/gircore/gir.core. But it's not production ready at the moment.

**Comment by BehroozAmoozad** [comment]
> Hi, I believe to understand Gtk4 would work on Windows. Is there any installer (msi/exe) available for that?
> 
> Let me know if it exists.

I've been using gtk4 on windows for 2 years. It works better than gtk3. msys2 works just fine.

Btw, anyone here considering helping out the gir.core project? that project is also getting a bit stale.
